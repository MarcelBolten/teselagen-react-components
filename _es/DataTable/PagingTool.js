var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp2;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from "react";
import { withProps, withHandlers, compose } from "recompose";
import classNames from "classnames";

import { noop, get } from "lodash";
import { Button, Classes } from "@blueprintjs/core";
import { pageSizes } from "./utils/queryParams";
import { onEnterOrBlurHelper } from "../utils/handlerHelpers";

export var PagingTool = (_temp2 = _class = function (_React$Component) {
  _inherits(PagingTool, _React$Component);

  function PagingTool() {
    var _temp, _this, _ret;

    _classCallCheck(this, PagingTool);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.state = {
      selectedPage: 1
    }, _this.onRefresh = function () {
      _this.props.onRefresh();
    }, _this.setPage = function (page) {
      _this.props.setPage(page);
      _this.props.onPageChange(page);
    }, _this.setPageSize = function (e) {
      _this.props.setPageSize(parseInt(e.target.value, 10));
    }, _this.pageBack = function () {
      var page = _this.props.paging.page;

      _this.setPage(parseInt(page, 10) - 1);
    }, _this.pageForward = function () {
      var page = _this.props.paging.page;

      _this.setPage(parseInt(page, 10) + 1);
    }, _this.setSelectedPage = function (e) {
      _this.setState({
        selectedPage: e.target.value
      });
    }, _this.pageInputBlur = function (e) {
      var _this$props$paging = _this.props.paging,
          pageSize = _this$props$paging.pageSize,
          total = _this$props$paging.total;

      var lastPage = Math.ceil(total / pageSize);
      var pageValue = parseInt(e.target.value, 10);
      var selectedPage = pageValue > lastPage ? lastPage : pageValue < 1 || isNaN(pageValue) ? 1 : pageValue;

      _this.setState({
        selectedPage: selectedPage
      });
      _this.setPage(selectedPage);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  PagingTool.prototype.UNSAFE_componentWillReceiveProps = function UNSAFE_componentWillReceiveProps(nextProps) {
    var page = nextProps.paging.page;

    this.setState({
      selectedPage: page
    });
  };

  PagingTool.prototype.render = function render() {
    var selectedPage = this.state.selectedPage;
    var _props = this.props,
        _props$paging = _props.paging,
        pageSize = _props$paging.pageSize,
        page = _props$paging.page,
        total = _props$paging.total,
        onRefresh = _props.onRefresh,
        disabled = _props.disabled;

    var pageStart = (page - 1) * pageSize + 1;
    if (pageStart < 0) throw new Error("We should never have page be <0");
    var backEnabled = page - 1 > 0;
    var forwardEnabled = page * pageSize < total;
    var lastPage = Math.ceil(total / pageSize);

    return React.createElement(
      "div",
      { className: "paging-toolbar-container" },
      onRefresh && React.createElement(Button, {
        minimal: true,
        icon: "refresh",
        disabled: disabled,
        onClick: this.onRefresh
      }),
      React.createElement(
        "div",
        { className: classNames(Classes.SELECT, Classes.MINIMAL) },
        React.createElement(
          "select",
          {
            className: "paging-page-size",
            onChange: this.setPageSize,
            disabled: disabled,
            value: pageSize
          },
          [React.createElement(
            "option",
            { key: "page-size-placeholder", disabled: true, value: "fake" },
            "Set Page Size"
          )].concat(pageSizes.map(function (size) {
            return React.createElement(
              "option",
              { key: size, value: size },
              size
            );
          }))
        )
      ),
      React.createElement(Button, {
        onClick: this.pageBack,
        disabled: !backEnabled || disabled,
        minimal: true,
        className: "paging-arrow-left",
        icon: "chevron-left"
      }),
      React.createElement(
        "div",
        null,
        total ? React.createElement(
          "div",
          null,
          React.createElement("input", _extends({
            style: { marginLeft: 5, width: 35, marginRight: 8 },
            value: selectedPage,
            disabled: disabled,
            onChange: this.setSelectedPage
          }, onEnterOrBlurHelper(this.pageInputBlur), {
            className: Classes.INPUT
          })),
          "of ",
          lastPage
        ) : "No Rows"
      ),
      React.createElement(Button, {
        style: { marginLeft: 5 },
        disabled: !forwardEnabled || disabled,
        icon: "chevron-right",
        minimal: true,
        className: "paging-arrow-right",
        onClick: this.pageForward
      })
    );
  };

  return PagingTool;
}(React.Component), _class.defaultProps = {
  onPageChange: noop
}, _temp2);

var ConnectedPagingTool = compose(withProps(function (props) {
  var entityCount = props.entityCount,
      page = props.page,
      pageSize = props.pageSize,
      disabled = props.disabled,
      onRefresh = props.onRefresh,
      setPage = props.setPage,
      setPageSize = props.setPageSize;

  return {
    paging: {
      total: entityCount,
      page: page,
      pageSize: pageSize
    },
    disabled: disabled,
    onRefresh: onRefresh,
    setPage: setPage,
    setPageSize: setPageSize
  };
}), withHandlers({
  onPageChange: function onPageChange(_ref) {
    var entities = _ref.entities,
        reduxFormSelectedEntityIdMap = _ref.reduxFormSelectedEntityIdMap;
    return function () {
      var record = get(entities, "[0]");
      if (!record || !record.id && record.id !== 0 && !record.code) {
        reduxFormSelectedEntityIdMap.input.onChange({});
      }
    };
  }
}))(PagingTool);

export default ConnectedPagingTool;