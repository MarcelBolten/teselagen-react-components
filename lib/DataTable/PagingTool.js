"use strict";

exports.__esModule = true;
exports.PagingTool = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _class, _temp2;

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _recompose = require("recompose");

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _lodash = require("lodash");

var _core = require("@blueprintjs/core");

var _queryParams = require("./utils/queryParams");

var _handlerHelpers = require("../utils/handlerHelpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PagingTool = exports.PagingTool = (_temp2 = _class = function (_React$Component) {
  _inherits(PagingTool, _React$Component);

  function PagingTool() {
    var _temp, _this, _ret;

    _classCallCheck(this, PagingTool);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.state = {
      selectedPage: 1
    }, _this.onRefresh = function () {
      _this.props.onRefresh();
    }, _this.setPage = function (page) {
      _this.props.setPage(page);
      _this.props.onPageChange(page);
    }, _this.setPageSize = function (e) {
      _this.props.setPageSize(parseInt(e.target.value, 10));
    }, _this.pageBack = function () {
      var page = _this.props.paging.page;

      _this.setPage(parseInt(page, 10) - 1);
    }, _this.pageForward = function () {
      var page = _this.props.paging.page;

      _this.setPage(parseInt(page, 10) + 1);
    }, _this.setSelectedPage = function (e) {
      _this.setState({
        selectedPage: e.target.value
      });
    }, _this.pageInputBlur = function (e) {
      var _this$props$paging = _this.props.paging,
          pageSize = _this$props$paging.pageSize,
          total = _this$props$paging.total;

      var lastPage = Math.ceil(total / pageSize);
      var pageValue = parseInt(e.target.value, 10);
      var selectedPage = pageValue > lastPage ? lastPage : pageValue < 1 || isNaN(pageValue) ? 1 : pageValue;

      _this.setState({
        selectedPage: selectedPage
      });
      _this.setPage(selectedPage);
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  PagingTool.prototype.UNSAFE_componentWillReceiveProps = function UNSAFE_componentWillReceiveProps(nextProps) {
    var page = nextProps.paging.page;

    this.setState({
      selectedPage: page
    });
  };

  PagingTool.prototype.render = function render() {
    var selectedPage = this.state.selectedPage;
    var _props = this.props,
        _props$paging = _props.paging,
        pageSize = _props$paging.pageSize,
        page = _props$paging.page,
        total = _props$paging.total,
        onRefresh = _props.onRefresh,
        disabled = _props.disabled;

    var pageStart = (page - 1) * pageSize + 1;
    if (pageStart < 0) throw new Error("We should never have page be <0");
    var backEnabled = page - 1 > 0;
    var forwardEnabled = page * pageSize < total;
    var lastPage = Math.ceil(total / pageSize);

    return _react2.default.createElement(
      "div",
      { className: "paging-toolbar-container" },
      onRefresh && _react2.default.createElement(_core.Button, {
        minimal: true,
        icon: "refresh",
        disabled: disabled,
        onClick: this.onRefresh
      }),
      _react2.default.createElement(
        "div",
        { className: (0, _classnames2.default)(_core.Classes.SELECT, _core.Classes.MINIMAL) },
        _react2.default.createElement(
          "select",
          {
            className: "paging-page-size",
            onChange: this.setPageSize,
            disabled: disabled,
            value: pageSize
          },
          [_react2.default.createElement(
            "option",
            { key: "page-size-placeholder", disabled: true, value: "fake" },
            "Set Page Size"
          )].concat(_queryParams.pageSizes.map(function (size) {
            return _react2.default.createElement(
              "option",
              { key: size, value: size },
              size
            );
          }))
        )
      ),
      _react2.default.createElement(_core.Button, {
        onClick: this.pageBack,
        disabled: !backEnabled || disabled,
        minimal: true,
        className: "paging-arrow-left",
        icon: "chevron-left"
      }),
      _react2.default.createElement(
        "div",
        null,
        total ? _react2.default.createElement(
          "div",
          null,
          _react2.default.createElement("input", _extends({
            style: { marginLeft: 5, width: 35, marginRight: 8 },
            value: selectedPage,
            disabled: disabled,
            onChange: this.setSelectedPage
          }, (0, _handlerHelpers.onEnterOrBlurHelper)(this.pageInputBlur), {
            className: _core.Classes.INPUT
          })),
          "of ",
          lastPage
        ) : "No Rows"
      ),
      _react2.default.createElement(_core.Button, {
        style: { marginLeft: 5 },
        disabled: !forwardEnabled || disabled,
        icon: "chevron-right",
        minimal: true,
        className: "paging-arrow-right",
        onClick: this.pageForward
      })
    );
  };

  return PagingTool;
}(_react2.default.Component), _class.defaultProps = {
  onPageChange: _lodash.noop
}, _temp2);


var ConnectedPagingTool = (0, _recompose.compose)((0, _recompose.withProps)(function (props) {
  var entityCount = props.entityCount,
      page = props.page,
      pageSize = props.pageSize,
      disabled = props.disabled,
      onRefresh = props.onRefresh,
      setPage = props.setPage,
      setPageSize = props.setPageSize;

  return {
    paging: {
      total: entityCount,
      page: page,
      pageSize: pageSize
    },
    disabled: disabled,
    onRefresh: onRefresh,
    setPage: setPage,
    setPageSize: setPageSize
  };
}), (0, _recompose.withHandlers)({
  onPageChange: function onPageChange(_ref) {
    var entities = _ref.entities,
        reduxFormSelectedEntityIdMap = _ref.reduxFormSelectedEntityIdMap;
    return function () {
      var record = (0, _lodash.get)(entities, "[0]");
      if (!record || !record.id && record.id !== 0 && !record.code) {
        reduxFormSelectedEntityIdMap.input.onChange({});
      }
    };
  }
}))(PagingTool);

exports.default = ConnectedPagingTool;