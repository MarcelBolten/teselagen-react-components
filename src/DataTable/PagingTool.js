//@flow
import React from "react";
import type { Paging } from "../flow_types";
import { NumericInput, Button } from "@blueprintjs/core";
import { onEnterOrBlurHelper } from "../utils/handlerHelpers";

export default class PagingTool extends React.Component {
  props: {
    paging: Paging,
    setPageSize: Function,
    onRefresh?: Function,
    setPage: Function
  };

  constructor(props) {
    super(props);

    this.state = {
      value: props.paging.pageSize
    };
  }

  onValueChange = value => {
    this.setState({
      value
    });
  };

  render() {
    const {
      paging: { pageSize, page, total },
      setPageSize,
      setPage,
      onRefresh
    } = this.props;
    const pageStart = (page - 1) * pageSize + 1;
    if (pageStart < 0) debugger;
    const pageEnd = (page - 1) * pageSize + pageSize < total
      ? (page - 1) * pageSize + pageSize
      : total;
    const backEnabled = page - 1 > 0;
    const forwardEnabled = page + 1 < (total + pageSize) / pageSize;
    return (
      <div className={"paging-toolbar-container"}>
        <span> Rows per page: </span>
        <NumericInput
          className={"paging-row-input"}
          value={this.state.value}
          onValueChange={value => {
            this.onValueChange(value);
            if (Math.abs(this.state.value - value) <= 1) {
              setPageSize(parseInt(value, 10));
            }
          }}
          {...onEnterOrBlurHelper(e => {
            this.onValueChange(e.target.value);
            setPageSize(parseInt(e.target.value, 10));
          })}
        />
        {onRefresh
          ? <Button iconName="refresh" onClick={() => onRefresh()} />
          : ""}
        <span
          onClick={() => {
            if (backEnabled) {
              setPage(page - 1);
            } else {
              toastr && toastr.warning("No more pages that way");
            }
          }}
          className={
            "pt-icon-standard pt-icon-arrow-left paging-arrow-left " +
              (!backEnabled && " pt-disabled")
          }
        />

        <span> {pageStart}-{pageEnd} of {total} </span>
        <span
          onClick={() => {
            if (forwardEnabled) {
              setPage(page + 1);
            } else {
              toastr && toastr.warning("No more pages that way");
            }
          }}
          className={
            "pt-icon-standard pt-icon-arrow-right paging-arrow-right " +
              (!forwardEnabled && " pt-disabled")
          }
        />

      </div>
    );
  }
  handleValueChange = (valueAsNumber: number, valueAsString: string) => {
    // console.log("Value as number:", valueAsNumber);
    // console.log("Value as string:", valueAsString);
  };
}
