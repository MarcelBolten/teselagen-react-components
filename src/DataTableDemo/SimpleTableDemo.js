import React from "react";
import { Button } from "@blueprintjs/core";
import { DataTable } from "../../../src";
import "./style.css";

const schema = {
  fields: [
    { path: "name" },
    {
      path: "id",
      type: "action",
      render: () => {
        return <Button className={"pt-minimal pt-icon-circle"} />;
      }
    }
  ]
};

const entities = [
  {
    name: "Thomas",
    id: "1"
  },
  {
    name: "Taoh",
    id: "2"
  },
  {
    name: "Chris",
    id: "3"
  },
  {
    name: "Sam",
    id: "4"
  },
  {
    name: "Adam",
    id: "5"
  }
];

export default function SimpleTable() {
  return (
    <div>
      <DataTable isSimple entities={entities} schema={schema} />
    </div>
  );
}
