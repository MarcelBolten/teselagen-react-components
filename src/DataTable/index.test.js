import React from "react";
import { shallow, mount } from "enzyme";
import Component from "./index";
import PagingTool from "./PagingTool";
import DataTableDemo, {
  UrlConnected,
  ReduxConnected,
  DataTableWrapper
} from "../../demo/src/DataTableDemo";

describe("(Component)", () => {
  it("renders without exploding", () => {
    const wrapper = shallow(<Component />);
    expect(wrapper).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders a search bar", () => {
    const clearFilters = jest.fn();
    const wrapper = shallow(
      <Component searchTerm="woeif" withSearch clearFilters={clearFilters} />
    );
    const clearFilterBtn = wrapper.find(".data-table-clear-filters");
    expect(clearFilterBtn).toHaveLength(1);
  });
});

describe("DataTableDemo", () => {
  const wrapper = mount(<DataTableDemo />);
  const urlWrapper = wrapper.find(UrlConnected);
  const reduxWrapper = wrapper.find(ReduxConnected);

  it("renders url and redux connected", () => {
    expect(urlWrapper).toHaveLength(1);
    expect(reduxWrapper).toHaveLength(1);
  });

  const urlDataWrapper = urlWrapper.find(DataTableWrapper);
  const reduxDataWrapper = reduxWrapper.find(DataTableWrapper);

  it("renders the datatable wrappers", () => {
    expect(urlDataWrapper).toHaveLength(1);
    expect(reduxDataWrapper).toHaveLength(1);
  });

  it("updates search term params on input", () => {
    const input = reduxDataWrapper.find(".datatable-search-input");
    input
      .find("input")
      .simulate("change", { target: { value: "search string" } });
    const button = input.find(".pt-icon-search");
    button.simulate("click");
    expect(reduxDataWrapper.props().tableParams.searchTerm).toEqual(
      "search string"
    );
    expect(reduxDataWrapper.props().queryParams).toEqual({
      include: {},
      limit: 10,
      offset: 0,
      order: "",
      where: {
        $or: {
          addedBy: { iLike: "%search string%" },
          name: { iLike: "%search string%" }
        }
      }
    });
  });

  describe("PagingTool", () => {
    const pagingToolbarWrapper = reduxDataWrapper.find(PagingTool);

    it("renders a PagingTool", () => {
      expect(pagingToolbarWrapper).toHaveLength(1);
    });

    const pagingInput = pagingToolbarWrapper.find("input");
    it("pagesize is initialized with a value of 10", () => {
      expect(pagingInput.props().value).toBe("10");
    });

    it("handles a page size change", () => {
      pagingInput.simulate("change", {
        target: {
          value: 2
        }
      });
      expect(pagingInput.props().value).toBe("2");
      pagingInput.simulate("blur");
      expect(reduxDataWrapper.props().queryParams.limit).toBe(2);
    });

    it("handles a page right", () => {
      pagingToolbarWrapper.find(".paging-arrow-right").simulate("click");
      expect(reduxDataWrapper.props().queryParams.offset).toBe(10);
      console.log(
        "reduxDataWrapper.props().queryParams",
        reduxDataWrapper.props().queryParams
      );
    });

    it("handles a page left after a page right", () => {
      expect(reduxDataWrapper.props().queryParams.offset).toBe(10);
      pagingToolbarWrapper.find(".paging-arrow-left").simulate("click");
      expect(reduxDataWrapper.props().queryParams.offset).toBe(0);
    });
  });
});
