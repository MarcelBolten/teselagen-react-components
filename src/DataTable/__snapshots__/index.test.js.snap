// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) renders without exploding 1`] = `
<div
  className="data-table-container "
>
  <div
    className="data-table-header"
  >
    <div
      className="data-table-title-and-buttons"
    />
    <div
      className="data-table-search-and-clear-filter-container"
    >
      <SearchBar
        reduxFormSearchInput={Object {}}
        setSearchTerm={[Function]}
      />
    </div>
  </div>
  <div
    className="data-table-body"
  >
    <Measure
      blacklist={Array []}
      cloneOptions={Object {}}
      includeMargin={true}
      onMeasure={[Function]}
      shouldMeasure={true}
      useClone={false}
      whitelist={
        Array [
          "width",
          "height",
          "top",
          "right",
          "bottom",
          "left",
        ]
      }
    >
      <Table
        allowMultipleSelection={true}
        columnWidths={Array []}
        defaultColumnWidth={150}
        defaultRowHeight={36}
        enableFocus={false}
        fillBodyWithGhostCells={true}
        isColumnResizable={false}
        isRowHeaderShown={false}
        loadingOptions={Array []}
        minColumnWidth={50}
        minRowHeight={20}
        numRows={10}
        onSelection={[Function]}
        renderBodyContextMenu={[Function]}
        renderRowHeader={[Function]}
        selectedRegionTransform={[Function]}
        selectedRegions={Array []}
        selectionModes={
          Array [
            3,
            2,
            1,
            0,
          ]
        }
      />
    </Measure>
  </div>
  <div
    className="data-table-footer"
  >
    <PagingTool
      paging={
        Object {
          "page": 0,
          "pageSize": 10,
          "total": undefined,
        }
      }
      setPage={[Function]}
      setPageSize={[Function]}
    />
  </div>
</div>
`;
